IGNITE-22829: Sql. Optimize performance of count(*)

lowka opened the pull request:
Optimizes:

- COUNT(*)
- COUNT(1)

No optimisation for COUNT(not_null_col)

```
SqlBenchmark.countKey                              1    false  avgt   20   0.142 ± 0.008  ms/op
SqlBenchmark.countVal                              1    false  avgt   20  22.455 ± 0.566  ms/op
```

```
Benchmark                      (clusterSize)  (fsync)  (tableSize)  Mode  Cnt   Score    Error  Units
SqlCountBenchmark.countNonOpt              1    false        30000  avgt   20  19.970 ±  0.138  ms/op
SqlCountBenchmark.countNonOpt              1    false        60000  avgt   20  38.292 ±  0.474  ms/op
SqlCountBenchmark.countNonOpt              1    false        90000  avgt   20  57.555 ±  0.261  ms/op
SqlCountBenchmark.countNonOpt              2    false        30000  avgt   20  12.183 ±  0.127  ms/op
SqlCountBenchmark.countNonOpt              2    false        60000  avgt   20  22.914 ±  0.299  ms/op
SqlCountBenchmark.countNonOpt              2    false        90000  avgt   20  34.088 ±  0.337  ms/op
SqlCountBenchmark.countNonOpt              3    false        30000  avgt   20   8.607 ±  0.142  ms/op
SqlCountBenchmark.countNonOpt              3    false        60000  avgt   20  16.460 ±  0.160  ms/op
SqlCountBenchmark.countNonOpt              3    false        90000  avgt   20  23.839 ±  0.304  ms/op
SqlCountBenchmark.countNonOpt              4    false        30000  avgt   20   9.920 ±  0.084  ms/op
SqlCountBenchmark.countNonOpt              4    false        60000  avgt   20  18.586 ±  0.243  ms/op
SqlCountBenchmark.countNonOpt              4    false        90000  avgt   20  26.447 ±  0.175  ms/op
SqlCountBenchmark.countNonOpt              5    false        30000  avgt   20   7.558 ±  0.148  ms/op
SqlCountBenchmark.countNonOpt              5    false        60000  avgt   20  14.384 ±  0.193  ms/op
SqlCountBenchmark.countNonOpt              5    false        90000  avgt   20  20.058 ±  0.143  ms/op
SqlCountBenchmark.countOpt                 1    false        30000  avgt   20   0.146 ±  0.001  ms/op
SqlCountBenchmark.countOpt                 1    false        60000  avgt   20   0.153 ±  0.001  ms/op
SqlCountBenchmark.countOpt                 1    false        90000  avgt   20   0.147 ±  0.003  ms/op
SqlCountBenchmark.countOpt                 2    false        30000  avgt   20   0.344 ±  0.007  ms/op
SqlCountBenchmark.countOpt                 2    false        60000  avgt   20   0.343 ±  0.007  ms/op
SqlCountBenchmark.countOpt                 2    false        90000  avgt   20   0.343 ±  0.006  ms/op
SqlCountBenchmark.countOpt                 3    false        30000  avgt   20   0.454 ±  0.011  ms/op
SqlCountBenchmark.countOpt                 3    false        60000  avgt   20   0.460 ±  0.013  ms/op
SqlCountBenchmark.countOpt                 3    false        90000  avgt   20   0.494 ±  0.017  ms/op
SqlCountBenchmark.countOpt                 4    false        30000  avgt   20   0.614 ±  0.021  ms/op
SqlCountBenchmark.countOpt                 4    false        60000  avgt   20   0.604 ±  0.017  ms/op
SqlCountBenchmark.countOpt                 4    false        90000  avgt   20   0.580 ±  0.034  ms/op
SqlCountBenchmark.countOpt                 5    false        30000  avgt   20   0.762 ±  0.018  ms/op
SqlCountBenchmark.countOpt                 5    false        60000  avgt   20   0.737 ±  0.022  ms/op
SqlCountBenchmark.countOpt                 5    false        90000  avgt   20   0.760 ±  0.031  ms/op
```


https://issues.apache.org/jira/browse/IGNITE-22829

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

AMashenkov commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:
```diff
@@ -238,26 +240,90 @@ public CompletableFuture<QueryPlan> prepareAsync(
+
+        IgniteRel fastOptRel = PlannerHelper.tryOptimizeSelectCount(planningContext.planner(), txContext, stmt.value.sqlNode());
+        if (fastOptRel == null) {
+            return new FastOptimizationResult(stmt, null);
```
Why do you need a wrapper FastOptimizationResult?
This method do not change `stmt` reference.

lowka replied:
Please take a look at how explain works.

korlov42 commented on modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/SelectCountPlannerTest.java:
```diff
@@ -0,0 +1,366 @@
+        QueryPlan plan = node.prepare("SELECT count(*) FROM test");
+
+        assertThat(plan, instanceOf(SelectCountPlan.class));
+        assertExpressions((SelectCountPlan) plan, "$0");
```
what does it mean?

lowka commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/IgniteSelectCount.java:
```diff
@@ -0,0 +1,131 @@
+            RelOptCluster cluster,
+            RelTraitSet traits,
+            @Nullable RelOptTable table,
+            RelDataType rowType,
```
Fixed.

lowka commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/IgniteSelectCount.java:
```diff
@@ -0,0 +1,131 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Relational operator that represents a SELECT COUNT(*) query.
```
Updated javadoc.

ygerzhedovich commented on modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSqlUsesSelectCountOptimizedTest.java:
```diff
@@ -0,0 +1,151 @@
+    public void countOptNoTable() {
+        assertQuery("SELECT COUNT(*)")
+                .matches(QueryChecker.containsSubPlan("SelectCount"))
+                .returns(1L)
```
I see that at least PG and MariaDB have the same result. But why? What does the construction mean?

lowka replied:
This query always produces a single row, thus count(1) returns 1.
But this code is removed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
+        // SELECT COUNT(*) ... row type
+        RelDataType countResultType = typeFactory.createSqlType(SqlTypeName.BIGINT);
+
+        RelDataTypeFactory.Builder inputRowBuilder = new Builder(typeFactory);
+        inputRowBuilder.add("ROWCOUNT", countResultType);
+        RelDataType inputRowType = inputRowBuilder.build();
```
looks like these lines can be safely removed: you already have created `countResultType` at line 422

lowka replied:
Thanks fixed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
+                return null;
+            }
+        } else {
+            // Try to optimize SELECT count(*) as well
```
what is the point of such optimization? What is the difference between new path and old path in numbers?

lowka replied:
Removed this piece of code.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
+            return null;
+        }
+
+        SqlSelect select = (SqlSelect) node;
```
```suggestion
        SqlSelect select = (SqlSelect) node;
        
        // make sure that the following IF statement does not leave out any operand of the SELECT node
        assert select.getOperandList().size() == 12 : "Expected 12 operands, but was " + select.getOperandList().size();

```

lowka replied:
Thanks. Fixed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
+                expressions.add(countValRef);
+
+                countAdded = true;
+            } else if (isCountNull(expr)) {
```
it's better to introduce general purpose optimization (under separate ticket), which will reduce `COUNT(NULL)` to simple `Values([[0]])` node.

lowka replied:
Removed this piece of code.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
+            if (isCountStar(planner.validator(), expr)) {
+                SqlCall call = (SqlCall) expr;
+
+                // Reject COUNT(DISTINCT id), but allow COUNT(DISTINCT 1)
```
`COUNT(DISTINCT 1)` must return 0 for empty table and 1 for non empty one, but after this patch latter case return N where N is a number of rows in a table.

lowka replied:
Thanks. Fixed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +356,145 @@ static boolean hasSubQuery(SqlNode node) {
+                || select.getHaving() != null
+                || select.getQualify() != null
+                || !select.getWindowList().isEmpty()
+                || select.getOrderList() != null && !select.getOrderList().isEmpty()
```
I see you've added orderList here. What is wrong with `ORDER BY`? I mean, at first glance ordering doesn't affect result of COUNT. Do I miss something?

lowka replied:
Thanks. Fixed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/SelectCountPlan.java:
```diff
@@ -0,0 +1,184 @@
+        IgniteTable igniteTable = optTable.unwrap(IgniteTable.class);
+        assert igniteTable != null;
+
+        CompletableFuture<Long> countFut  = tableRegistry.getTable(catalogVersion, igniteTable.id())
```
```suggestion
        CompletableFuture<Long> countFut = tableRegistry.getTable(catalogVersion, igniteTable.id())
```

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/IgniteSelectCount.java:
```diff
@@ -0,0 +1,150 @@
+    @Override
+    public RelWriter explainTerms(RelWriter pw) {
+        RelWriter relWriter = super.explainTerms(pw);
+        if (table != null) {
```
looks like this `if` statement is not necessary anymore

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:
```diff
@@ -238,26 +241,102 @@ public CompletableFuture<QueryPlan> prepareAsync(
+        CompletableFuture<ValidStatement> validFut = validateStatement(parsedResult, planningContext);
+
+        // Try optimize to fast
+        CompletableFuture<FastOptimizationResult> plannedOrValidated =
```
tbh, I didn't get the idea behind introduction of additional container for composite result.

it's better to do this:
```
        CompletableFuture<ValidStatement> validFut = validateStatement(parsedResult, planningContext);

        result = validFut.thenCompose(stmt -> {
            QueryPlan plan = tryOptimizeFast(stmt, planningContext, txContext);

            // If optimize fast returned a plan, return it.
            if (plan != null) {
                return CompletableFuture.completedFuture(plan);
            }

            // Otherwise, continue with the regular planning.
            return prepareAsync0(stmt, planningContext, txContext);
        });
```

lowka replied:
Fixed.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:
```diff
@@ -295,19 +379,36 @@ private CompletableFuture<QueryPlan> prepareExplain(ParsedResult parsedResult, P
+
+        // Try optimize to fast
+        CompletableFuture<FastOptimizationResult> plannedOrValidated =
+                validFut.thenApply(stmt -> tryOptimizeFast(stmt, ctx, txContext));
```
optimization flow during EXPLAIN and regular flow must share the same path.

lowka replied:
Fixed.

korlov42 reviewed (changes requested):
Unfortunately, this patch introduces significant performance degradation for other cases:

```
Before (main at 513d54d6)
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt   Score   Error  Units
InsertBenchmark.sqlPreparedInsert              1    false                25  avgt   20  66.751 ± 5.146  us/op

After
Benchmark                          (clusterSize)  (fsync)  (partitionCount)  Mode  Cnt    Score   Error  Units
InsertBenchmark.sqlPreparedInsert              1    false                25  avgt   20  119.580 ± 7.993  us/op
```

lowka commented:
@korlov42 Thanks, fixed. 
Removed an unconditional call to sql validator that always bypassed caching.
```
main
InsertBenchmark.sqlPreparedInsert              1    false                 1  avgt   10  70.827 ± 2.993  us/op

branch
InsertBenchmark.sqlPreparedInsert              1    false                 1  avgt   10  71.370 ± 5.437  us/op
```

lowka commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +357,228 @@ static boolean hasSubQuery(SqlNode node) {
+
+    /**
+     * Whether we can optimize a query that looks like {@code SELECT count(*)} or not.
+     * If this method returns {@code true}, then {@link #tryOptimizeSelectCount(IgnitePlanner, QueryTransactionContext, SqlNode)}
```
Fixed. This method is removed as it is no longer needed.

lowka commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:
```diff
@@ -248,16 +252,20 @@ public CompletableFuture<QueryPlan> prepareAsync(
+    private CompletableFuture<QueryPlan> prepareAsync0(
+            ParsedResult parsedResult,
+            PlanningContext planningContext,
+            @Nullable QueryTransactionContext txContext
```
Fixed.

lowka commented on modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSqlUsesSelectCountOptimizedTest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Test
+    public void countOpt() {
```
Added column name checks.

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java:
```diff
@@ -343,4 +355,164 @@ static boolean hasSubQuery(SqlNode node) {
+        }
+    }
+
+    private static boolean isCountStartOperand(SqlNode node) {
```
name of this method is quite confusing. Why `NULL` is countStart operand while any literal is not?

lowka replied:
Fixed. Moved the body of this method to `isCountStar`.
